internals/src/compact_size.rs

```diff
@@ -0,0 +1,232 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Variable length integer encoding A.K.A [`CompactSize`].
+//!
+//! An integer can be encoded depending on the represented value to save space. Variable length
+//! integers always precede an array/vector of a type of data that may vary in length.
+//!
+//! [`CompactSize`]: <https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer>
+
+use crate::array_vec::ArrayVec;
+
+/// The maximum size of a serialized object in bytes or number of elements
+/// (for eg vectors) when the size is encoded as `CompactSize`.
+///
+/// This is `MAX_SIZE` in Bitcoin Core.
+// Issue: https://github.com/rust-bitcoin/rust-bitcoin/issues/3264
+pub const MAX_ENCODABLE_VALUE: u64 = 0x0200_0000;
+
+/// The maximum length of an encoding.
+const MAX_ENCODING_SIZE: usize = 9;
+
+/// Returns the number of bytes used to encode this `CompactSize` value.
+///
+/// # Returns
+///
+/// - 1 for 0..=0xFC
+/// - 3 for 0xFD..=(2^16-1)
+/// - 5 for 0x10000..=(2^32-1)
+/// - 9 otherwise.
+#[inline]
+pub const fn encoded_size(value: u64) -> usize {
```

Comment:
Could've been `ToU64` to simplify the callers.

Reply:
Ha! Can't call that in `const`, sadly I had to make the change before I remembered why I didn't use it originally :)

Reply:
const trait methods would be awesome.

Reply:
I just removed the `const` - we will likely need to add it again later but in this PR its better as you suggest.